Node.js SDK : Error handling

benjaminstrasser opened the pull request:
- [x] SDK throws basic errors 
- [x] Create own centralized error handling with custom dagger errors 
  - [x] Replace basic errors with custom dagger SDK errors
  - [x] https://github.com/dagger/dagger/pull/3956#discussion_r1030628831 
  - [x] Others
- [x] Error documentation
  - [x] Documemntation via comments
  - [x] Add centralized error handling to docusaurus
- [x] ~~Optional: Amend documentation to adopt explicit error handling~~

benjaminstrasser commented:
Thanks @slumbering  for the input and feedback!
> What about creating a new file `ErrorHandler.ts` or whatever you feel to centralize error handling ? A even cleaner way would be to create one file per error type. `GraphRequestError.ts` `EngineConnectError.ts` and so forth...
> 
Sound like a good plan. I will give it a try after work.

> Finally users can get a precise error:
> 
> ```ts
> connect(async (client: Client) => {
>     ....
> }).catch((error) => {
>   // We can then access to the error message or code, depending of what we want
>    console.error(error.message);
>    console.error(error.code);
> });
> ```
Here the question remains if we want to update the guides to reflect this possibility. I would personally add it but don't have particularly strong feelings about this.


About error references is see two options:
- use docusaurus and f.e. just add `errors.ts` https://github.com/dagger/dagger/blob/64a40a222640842b2f57c2829fd905c1db416aa3/website/docusaurus.config.js#L136-L154
- Write docs by hand, like f.e. [Angular Error List](https://angular.io/errors)

I would personally prefer option one and create a more detailed error reference once the SDK matures a bit more, but am open for input on this.

slumbering commented:
> 



> Thanks @slumbering for the input and feedback!
> 
> > What about creating a new file `ErrorHandler.ts` or whatever you feel to centralize error handling ? A even cleaner way would be to create one file per error type. `GraphRequestError.ts` `EngineConnectError.ts` and so forth...
> 
> Sound like a good plan. I will give it a try after work.
> 
> > Finally users can get a precise error:
> > ```ts
> > connect(async (client: Client) => {
> >     ....
> > }).catch((error) => {
> >   // We can then access to the error message or code, depending of what we want
> >    console.error(error.message);
> >    console.error(error.code);
> > });
> > ```
> 
> Here the question remains if we want to update the guides to reflect this possibility. I would personally add it but don't have particularly strong feelings about this.
> 
> About error references is see two options:
> 
> * use docusaurus and f.e. just add `errors.ts` https://github.com/dagger/dagger/blob/64a40a222640842b2f57c2829fd905c1db416aa3/website/docusaurus.config.js#L136-L154
> * Write docs by hand, like f.e. [Angular Error List](https://angular.io/errors)
> 
> I would personally prefer option one and create a more detailed error reference once the SDK matures a bit more, but am open for input on this.

Agreed, option one is definitely what we want 👍, we do not want to maintain by hand ref docs.

benjaminstrasser commented:
@slumbering 
Any feedback on the last commit? would love some feedback before starting the docs.
Was sticking with SDK errors only for now. How to interpret GraphQL errors is a little out of scope for this PR imo, but should probably be addressed.

slumbering commented:
Awesome @benjaminstrasser thanks !
I'm going to have a look now 👀

slumbering commented on sdk/nodejs/common/errors.ts:
```diff
@@ -0,0 +1,184 @@
+*/
+export class GraphQLRequestError extends DaggerSDKError {
+  name = "GraphQLRequestError"
+  code = "100"
```
Let's prefix the code with an `D` like `D100` (D stands for Dagger)

slumbering commented:
I'm gonna push some changes in few minutes cc @benjaminstrasser

slumbering commented:
@benjaminstrasser ! I've just push a commit. Basically i've created a more atomic structure with one error per file. I've also installed `node-color-log` to improve error readability in the terminal (no particular preference for this library feel free to change it if there is one that you prefer).

benjaminstrasser commented:
@slumbering 
Hey. Thanks for the commit! I already started doing the basic typedoc documentation and will merge and push them shortly.
After that, imo we are ready for a final review.

benjaminstrasser commented:
Hey @slumbering 
I manually merged most of your and my changes by hand as the automatic merge from intellij was a mess and the manual one was very easy. This gets squished anyway so I hope you forgive me for hacking on this one ;D

I added basic documentation to all errors and added them to docusaurus to autogenerate references. I will welcome any additions and feedback to the comments I wrote. Some of the people who wrote the code that errors could probably elaborate on them a lot better.

I also removed the automatic logging. Imo there is no need to log errors regardless of what happens to them. F.e in our case the error was logged during the test even though it was caught, which I find a little confusing.
I would much rather have the user choose to pretty print them if they desire to do so. This is why I created a `printStackTrace`method they can use and maybe we can add to the examples and such. Maybe I am wrong on this tho. Would love your feedback as well

I think this is it for this PR and I would love a review.

slumbering commented:
Hey @benjaminstrasser ! I'm going to review this PR very soon :). We are going to ship some breaking changes this week and then the error handling can be next (next week or the week after). I let you know 👍

slumbering commented:
Hey @benjaminstrasser ! I just rebased the main branch. Many changes has been done and I've all the context. That's why I did this rebase. We should be ready today for merging :).

slumbering commented on sdk/nodejs/test/connect.spec.ts:
```diff
@@ -9,11 +10,28 @@ describe("NodeJS sdk", function () {
-        .withExec(["apk", "add", "curl"])
-        .withExec(["curl", "https://dagger.io/"])
-        .stdout()
+        .exec(["apk", "add", "curl"])
```
```suggestion
        .withExec(["apk", "add", "curl"])
```

slumbering commented on sdk/nodejs/common/errors/UnknownDaggerError.ts:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ *  This error is thrown if the dagger SDK does not identify the error and just wraps the cause.
+ *  @see [UnknownDaggerError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.UnknownDaggerError)
```
The `@see` param links the page to itself.

slumbering replied:
I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/DaggerSDKError.ts:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * The base error. Every other error inherits this error.
+ * @see [DaggerSDKError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.DaggerSDKError)
```
The `@see` param links the page to itself.
eg: https://deploy-preview-3947--devel-docs-dagger-io.netlify.app/current/sdk/nodejs/reference/classes/common_errors.DaggerSDKError

I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/DockerImageRefValidationError.ts:
```diff
@@ -0,0 +1,28 @@
+/**
+ *  This error is thrown if the passed image reference does not pass validation and is not compliant with the
+ *  DockerImage constructor.
+ *  @see [DockerImageRefValidationError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.DockerImageRefValidationError)
```
The `@see` param links the page to itself.
I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/EngineSessionPortParseError.ts:
```diff
@@ -0,0 +1,31 @@
+ * - Reading the port times out after 30 seconds
+ * - The parsed port is not a number
+ * - the reader does not read a single line
+ * @see [EngineSessionPortParseError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.EngineSessionPortParseError)
```
The `@see` param links the page to itself.

slumbering replied:
I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/GraphQLRequestError.ts:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ *  This error originates from the dagger engine. It means that some error was thrown and sent back via GraphQL.
+ *  @see [GraphQLRequestError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.GraphQLRequestError)
```
The `@see` param links the page to itself.

slumbering replied:
I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/InitEngineSessionBinaryError.ts:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ *  This error is thrown if the dagger binary cannot be copied from the dagger docker image and copied to the local host.
+ *  @see [InitEngineSessionBinaryError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.InitEngineSessionBinaryError)
```
The `@see` param links the page to itself.

slumbering replied:
I guess we should remove it.

slumbering commented on sdk/nodejs/common/errors/TooManyNestedObjectsError.ts:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ *  Dagger only expects one response value from the engine. If the engine returns more than one value this error is thrown.
+ *  @see [TooManyNestedObjectsError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.TooManyNestedObjectsError)
```
The `@see` param links the page to itself.

slumbering reviewed (commented):
Tiny comments, nothing crazy. Once we're good, let's ship it 
🚢 🚀

benjaminstrasser commented:
@slumbering 
Think I added all your feedback. Thanks for the review

slumbering commented:
LGTM ! 🚀
